### PR TITLE
Create mobile renderer

### PR DESCRIPTION
**Pull request by BastiaanOlij**
This PR adds a mobile version of the Vulkan renderer to Godot as the full clustered renderer uses a few techniques not supported on mobile hardware. Changes (will) include:
- no depth pre-pass
- normal light culling as we can't used the clustered approach
- no GI and likely no volumetric fog (maybe)
- move most things related to the mobile renderer into a subfolder (just like we did for clustered)
- instance information is now sent through the push constant instead of a bigger UBO

### CODE REVIEW

File: servers/rendering/renderer_rd/shaders/light_data_inc.glsl [hunk 1]
```diff
@@ -0,0 +1,83 @@
+
+struct LightData { //this structure needs to be as packed as possible
+	vec3 position;
+	float inv_radius;
```
Comment (fire) [inline servers/rendering/renderer_rd/shaders/light_data_inc.glsl:4]: Newer coding style is to initialize declarations.

**Comment by BastiaanOlij** [comment]
Got to a point today where it is starting to render stuff unshaded. Rendering of objects is now done through push constants. I do want to pretty up how the push constants are populated

Next step is to get lights to work again

**Comment by BastiaanOlij** [comment]
Slow progress today but directional lights seem to work and transparency now works

image

There are a few bugs that snuck in that I will look at over the weekend, for instance this is totally screwed up now:

image

Then I'm going to make the shadow render pass work.

I think I'll also make the background work first because it'll make things look much nicer :)

**Comment by BastiaanOlij** [comment]
Also currently still crashes when `directional_shadow_atlas_set_size` is called on exit. Not sure why:
image

**Comment by BastiaanOlij** [comment]
I put back most of the shadow code but it's failing to run it. I'm getting a bunch of errors that the shader variant is not setup:

```
ERROR: Attempted to use an unused shader variant (shader is null),
   at: (D:\Development\godot4-git\servers/rendering/renderer_rd/pipeline_cache_rd.h:71)
ERROR: Condition "!pipeline" is true.
   at: RenderingDeviceVulkan::draw_list_bind_render_pipeline (drivers\vulkan\rendering_device_vulkan.cpp:6379)
ERROR: The vertex format used to create the pipeline does not match the vertex format bound.
   at: (drivers\vulkan\rendering_device_vulkan.cpp:6586)
ERROR: Attempted to use an unused shader variant (shader is null),
   at: (D:\Development\godot4-git\servers/rendering/renderer_rd/pipeline_cache_rd.h:71)
ERROR: This render pipeline requires (0) bytes of push constant data, supplied: (128)
   at: (drivers\vulkan\rendering_device_vulkan.cpp:6562)
ERROR: No render pipeline was set before attempting to draw.
   at: (drivers\vulkan\rendering_device_vulkan.cpp:6579)
ERROR: Attempted to use an unused shader variant (shader is null),
   at: (D:\Development\godot4-git\servers/rendering/renderer_rd/pipeline_cache_rd.h:71)
ERROR: This render pipeline requires (0) bytes of push constant data, supplied: (128)
   at: (drivers\vulkan\rendering_device_vulkan.cpp:6562)
ERROR: No render pipeline was set before attempting to draw.
   at: (drivers\vulkan\rendering_device_vulkan.cpp:6579)
ERROR: Attempted to use an unused shader variant (shader is null),
   at: (D:\Development\godot4-git\servers/rendering/renderer_rd/pipeline_cache_rd.h:71)
ERROR: This render pipeline requires (0) bytes of push constant data, supplied: (128)
   at: (drivers\vulkan\rendering_device_vulkan.cpp:6562)
```

I think this is due to the way the shadow variants are parsed in `SceneShaderForwardClustered::ShaderData::set_code`, looks like it won't include them because we're not using the depth prepass but I could be misreading this code and thus not have changed the code correctly. Old code from the forward clustered shader is commented out in the same block.

**Comment by BastiaanOlij** [comment]
Have to commit the changes but the sky works:
image
Full kudos to @clayjohn though, the code was well separated out of the main engine so all that work I did moving it into its own class paid off. It was really easy to hang it back in there and make it work. 

I also have shadows!
I'm still getting the above mentioned errors, I think this is related to the material shader

File: servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp [hunk 2]
```diff
@@ -0,0 +1,2023 @@
...
+	rl->element_info.resize(p_offset + element_total);
+
+	uint32_t repeats = 0;
+	GeometryInstanceSurfaceDataCache *prev_surface = nullptr;
+	for (uint32_t i = 0; i < element_total; i++) {
+		GeometryInstanceSurfaceDataCache *surface = rl->elements[i + p_offset];
+		GeometryInstanceForwardMobile *inst = surface->owner;
+
+		/* 
+		// !BAS! moved this into _render_list_template, same amount of work but no memory wasted. Keeping commented out until we make a decision which way we'll go.
```
Comment (BastiaanOlij) [inline servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:1143]: @reduz so our original idea was to populate these and store them in our class but most of them change each frame so there doesn't seem to much sense to it. I've left the code commented out here just in case we change our mind but it now happens inside of _render_list_template. I think the way I do it now is fine but let me know if you agree.

**Comment by BastiaanOlij** [comment]
> 
> 
> Just wondering, will we be able to use "mobile" renderer on desktop (power saving on a laptop and/or bad Vulkan support)?

Yup, it'll just be a setting in project settings, similar to GLES2/GLES3 switch

**Comment by BastiaanOlij** [comment]
Today I managed to get all the lights, reflection probes and decals working. Reflection probes do have issues but those also are apparent in the clustered renderer.

image

Tomorrow I'm going to add msaa back in and try to finally figure out why one of the shader variants isn't building causing weird error spam.

**Comment by BastiaanOlij** [comment]
I've added MSAA support but there is an issue around the 16x version giving a black screen on my laptop that I still need to look into.

Multimesh seems to work
Bones seem to work. 
Both of these basically use the same approach as before loading the needed data into the transform UBO

Also less errors around the shaders so that's nearly working.

Agreed with Reduz that we're going to do further improvements on lightmaps in another PR. I do still want to test if it works (or @JFonS maybe you want to give my branch a quick twirl? In theory they should work)

Basically all thats left is cleanup I think.

**Comment by JFonS** [comment]
> Agreed with Reduz that we're going to do further improvements on lightmaps in another PR. I do still want to test if it works (or @JFonS maybe you want to give my branch a quick twirl? In theory they should work)

Lightmaps are not easy to test because unwrapping is not working right now. We can wait until I re-add it and then revisit lightmaps on the mobile renderer.

**Comment by Calinou** [comment]
> I've added MSAA support but there is an issue around the 16x version giving a black screen on my laptop that I still need to look into.

See https://github.com/godotengine/godot/issues/37917.

**Comment by BastiaanOlij** [comment]
> 
> 
> > Agreed with Reduz that we're going to do further improvements on lightmaps in another PR. I do still want to test if it works (or @JFonS maybe you want to give my branch a quick twirl? In theory they should work)
> 
> Lightmaps are not easy to test because unwrapping is not working right now. We can wait until I re-add it and then revisit lightmaps on the mobile renderer.

I think that's fine because even if the current logic doesn't work we'll have to redo it anyway. The problem as I understand it with the mobile renderer doesn't allow as large buffers to be loaded as on desktop which is also why we're now sending some of the code data right before the draw call.

The plan is to do the same with selecting the lightmap. By sorting our drawbuffer by lightmap we can change the current selected lightmap as few times as possible and access it during rendering, instead of looking up a lightmap without our lightmap buffer. 

But reduz already mentioned we should do this in a follow up PR.

**Comment by BastiaanOlij** [comment]
> 
> 
> > I've added MSAA support but there is an issue around the 16x version giving a black screen on my laptop that I still need to look into.
> 
> See #37917. Given the high cost of 16× MSAA, I wouldn't bother supporting it in 2021, especially in a renderer targeting mobile.

Buffer, well that explains it. Agreed it probably has no purpose in a mobile renderer but seeing its a global settings it's possible to set it. That may need to be resolved in a follow up PR or in the PR that fixes the MSAA :)

**Comment by clayjohn** [comment]
Early results look great! I am getting about half the frame time that I do with the default renderer with a very simple scene.

File: servers/rendering/renderer_rd/renderer_compositor_rd.cpp [hunk 3]
```diff
@@ -175,5 +175,20 @@ RendererCompositorRD::RendererCompositorRD() {
 
 	storage = memnew(RendererStorageRD);
 	canvas = memnew(RendererCanvasRenderRD(storage));
-	scene = memnew(RendererSceneRenderImplementation::RenderForwardClustered(storage));
+
+	// !BAS! should we move the definition somewhere else?
+	String mode = GLOBAL_DEF("rendering/vulkan/rendering/mode", "Forward clustered");
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/renderer_compositor_rd.cpp:180]: This should be moved to rendering_server.cpp here:
https://github.com/godotengine/godot/blob/0ee744ff5f3cf045a2699aa4111de3200b9d70be/servers/rendering_server.cpp#L2285

And it should replace ``use_low_end_renderer``

Also it should use ``GLOBAL_DEF_RST`` so that it suggests restarting the editor for the user

File: servers/rendering_server.cpp [hunk 4]
```diff
@@ -2297,8 +2297,12 @@ RenderingServer::RenderingServer() {
 
 	GLOBAL_DEF("rendering/2d/shadow_atlas/size", 2048);
 
-	GLOBAL_DEF("rendering/driver/rd_renderer/use_low_end_renderer", false);
-	GLOBAL_DEF("rendering/driver/rd_renderer/use_low_end_renderer.mobile", true);
+	GLOBAL_DEF_RST("rendering/vulkan/rendering/mode", "Forward clustered");
```
Comment (reduz) [inline servers/rendering_server.cpp:2300]: Dont use strings here, its kinda unsafe. use an enum (0,1). Here suggested change:

```C++
GLOBAL_DEF_RST("rendering/vulkan/rendering/back_end", 0);
GLOBAL_DEF_RST("rendering/vulkan/rendering/back_end.mobile", 1);

ProjectSettings::get_singleton()->set_custom_property_info("rendering/vulkan/rendering/back_end",
			PropertyInfo(Variant::STRING,
					"rendering/vulkan/rendering/back_end",
					PROPERTY_HINT_ENUM, "ForwardClustered,ForwardMobile"));
					
					```

**Review by reduz — CHANGES_REQUESTED** [review]

File: servers/rendering_server.cpp (hunk 4 above)
Comment (BastiaanOlij) [reply servers/rendering_server.cpp:2300]: Done :)

**Review by reduz — APPROVED** [review]

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

File: servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp [hunk 5]
```diff
@@ -0,0 +1,2163 @@
...
+}
+
+/* these are being called from RendererSceneRenderRD::_pre_opaque_render */
+
+void RenderForwardMobile::_render_shadow_begin() {
+	scene_state.shadow_passes.clear();
+	RD::get_singleton()->draw_command_begin_label("Shadow Setup");
+	_update_render_base_uniform_set();
+
+	render_list[RENDER_LIST_SECONDARY].clear();
```
Comment (jamie-pate) [inline servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:559]: https://github.com/BastiaanOlij/godot/blob/58ff0dac1a3ff800128582575f2b9b1b95e40962/servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp#L1505

This instance_data doesn't need to be cleared in mobile?
`scene_state.instance_data[RENDER_LIST_SECONDARY].clear();`

Comment (jamie-pate) [reply servers/rendering/renderer_rd/forward_mobile/render_forward_mobile.cpp:559]: NVM: updated in 4.5 https://github.com/godotengine/godot/pull/103641/files
